File: frontend/src/lib/access.ts
import type { User } from './auth';

/** Effective permission codenames from `GET /auth/me/` — mirror `backend/core/access.py`. */
export const P = {
  SUPER_METRICS: 'feature.super_metrics',
  POS_ADMIN: 'feature.pos_admin',
  POS_RETAIL_LANE: 'feature.pos_retail_lane',
  POS_WHOLESALE: 'feature.pos_wholesale',
  POS_WHOLESALE_ADMIN: 'feature.pos_wholesale_admin',
  INVOICE_ADMIN_STORES: 'feature.invoice_admin_stores',
  INVOICE_RESTRICTED: 'feature.invoice_restricted',
  INVOICE_HIDE_CASH_CHECKOUT: 'feature.invoice_hide_cash_checkout',
  RETAIL_CATALOG_RESTRICTED: 'feature.retail_catalog_restricted',
  LEDGER_ADMIN: 'feature.ledger_admin',
  STORE_MANAGEMENT: 'feature.store_management',
  PAYMENTS_EXTENDED: 'feature.payments_extended_columns',
  DISCARD_INVOICE_EDIT_CARTS: 'feature.discard_invoice_edit_carts',
  ROLE_MANAGEMENT: 'feature.role_management',
} as const;

function normalizedGroups(user: User | null | undefined): Set<string> {
  return new Set((user?.groups || []).map((g) => String(g || '').trim().toLowerCase()));
}

export function hasAdminSuperBypass(user: User | null | undefined): boolean {
  const groups = normalizedGroups(user);
  return groups.has('admin') && groups.has('super');
}

export function hasPermission(
  user: User | null | undefined,
  codename: string,
  legacy: () => boolean
): boolean {
  // Group-first priority: explicit Admin + Super bypass.
  if (hasAdminSuperBypass(user)) return true;
  if (user?.is_superuser || user?.is_staff) return true;
  const p = user?.permissions;
  if (Array.isArray(p)) return p.includes(codename);
  return legacy();
}

export function hasNavPermission(user: User | null | undefined, navCodename: string): boolean {
  return hasPermission(user, navCodename, () => false);
}

export function isPosAdminContext(user: User | null | undefined): boolean {
  return hasPermission(user, P.POS_ADMIN, () => Boolean(user?.is_admin || user?.is_superuser || user?.is_staff));
}

export function isPosRetailLane(user: User | null | undefined): boolean {
  return hasPermission(user, P.POS_RETAIL_LANE, () => false);
}

export function isPosWholesaleLane(user: User | null | undefined): boolean {
  if (hasAdminSuperBypass(user)) return false;
  return hasPermission(user, P.POS_WHOLESALE, () => false);
}

export function isPosWholesaleAdmin(user: User | null | undefined): boolean {
  return hasPermission(user, P.POS_WHOLESALE_ADMIN, () => false);
}

export function isPosWholesaleStaffOnly(user: User | null | undefined): boolean {
  return isPosWholesaleLane(user) && !isPosWholesaleAdmin(user);
}

export function isInvoiceAdminStores(user: User | null | undefined): boolean {
  return hasPermission(user, P.INVOICE_ADMIN_STORES, () => false);
}

export function canSeeSuperMetrics(user: User | null | undefined): boolean {
  return hasPermission(user, P.SUPER_METRICS, () => false);
}

export function isInvoiceRestrictedUser(user: User | null | undefined): boolean {
  if (hasAdminSuperBypass(user)) return false;
  return hasPermission(user, P.INVOICE_RESTRICTED, () => false);
}

export function hasInvoiceHideCashCheckout(user: User | null | undefined): boolean {
  if (hasAdminSuperBypass(user)) return false;
  return hasPermission(user, P.INVOICE_HIDE_CASH_CHECKOUT, () => false);
}

export function isRetailCatalogRestricted(user: User | null | undefined): boolean {
  if (hasAdminSuperBypass(user)) return false;
  return hasPermission(user, P.RETAIL_CATALOG_RESTRICTED, () => false);
}

export function isLedgerAdminContext(user: User | null | undefined): boolean {
  return hasPermission(user, P.LEDGER_ADMIN, () => Boolean(user?.is_admin || user?.is_superuser || user?.is_staff));
}

export function isStoreManagementAdmin(user: User | null | undefined): boolean {
  return hasPermission(user, P.STORE_MANAGEMENT, () => Boolean(user?.is_admin || user?.is_staff || user?.is_superuser));
}

export function hasPaymentsExtendedColumns(user: User | null | undefined): boolean {
  return hasPermission(user, P.PAYMENTS_EXTENDED, () => false);
}

export function canDiscardInvoiceEditCarts(user: User | null | undefined): boolean {
  return hasPermission(user, P.DISCARD_INVOICE_EDIT_CARTS, () => false);
}

export function canManageRoles(user: User | null | undefined): boolean {
  return hasPermission(user, P.ROLE_MANAGEMENT, () => false);
}

File: backend/core/access.py
"""
Navigation / feature permission codenames.

SaaS model: app access is role-driven via Role + UserStoreRole. Django groups may
still exist for admin/platform workflows, but they should not encode tenant app ACLs.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from backend.core.models import User

# All UI routes and feature gates — keep in sync with Layout.tsx `permission` fields.
ACCESS_PERMISSION_SEED: list[tuple[str, str, str]] = [
    ('nav.pos', 'POS', 'core'),
    ('nav.repair_register', 'New repair', 'core'),
    ('nav.search', 'Search', 'core'),
    ('nav.dashboard', 'Dashboard', 'core'),
    ('nav.invoices', 'Invoices', 'sales'),
    ('nav.credit_notes', 'Credit notes', 'sales'),
    ('nav.customers', 'Customers', 'sales'),
    ('nav.replacement', 'Replacement', 'sales'),
    ('nav.repairs', 'Repairs', 'sales'),
    ('nav.products', 'Products', 'inventory'),
    ('nav.stock_overview', 'Stock overview', 'inventory'),
    ('nav.stock_transfers', 'Stock transfers', 'inventory'),
    ('nav.purchases', 'Purchases', 'inventory'),
    ('nav.ledger', 'Ledger', 'financial'),
    ('nav.personal_ledger', 'Personal ledger', 'financial'),
    ('nav.internal_ledger', 'Internal / shop ledger', 'financial'),
    ('nav.payment_reminders', 'Payment reminders', 'financial'),
    ('nav.expenses', 'Expenses', 'financial'),
    ('nav.payments', 'Payments', 'financial'),
    ('nav.active_carts', 'Active carts', 'admin'),
    ('nav.vendors', 'Vendors', 'admin'),
    ('nav.reports', 'Reports', 'admin'),
    ('nav.history', 'History', 'admin'),
    ('nav.role_management', 'Role management', 'admin'),
]

# Page-level gates (POS, invoices, ledger, etc.) — keep in sync with frontend `lib/access.ts` `P`.
FEATURE_PERMISSION_SEED: list[tuple[str, str, str]] = [
    ('feature.super_metrics', 'Super-only metrics & listings', 'feature'),
    ('feature.pos_admin', 'POS / repair global admin lane', 'feature'),
    ('feature.pos_retail_lane', 'POS retail lane (store selector)', 'feature'),
    ('feature.pos_wholesale', 'POS wholesale lane', 'feature'),
    ('feature.pos_wholesale_admin', 'POS wholesale admin (store selector)', 'feature'),
    ('feature.invoice_admin_stores', 'Invoices: all stores / admin store UI', 'feature'),
    ('feature.invoice_restricted', 'Invoice detail: restricted retail/wholesale editor', 'feature'),
    ('feature.invoice_hide_cash_checkout', 'Hide cash/UPI checkout options (wholesale)', 'feature'),
    ('feature.retail_catalog_restricted', 'Products/purchases: retail-only restrictions', 'feature'),
    ('feature.ledger_admin', 'Ledger: admin lane (substring Admin in group)', 'feature'),
    ('feature.store_management', 'Stores CRUD: admin / store admins', 'feature'),
    ('feature.payments_extended_columns', 'Payments: extended columns (non-Retail group)', 'feature'),
    ('feature.discard_invoice_edit_carts', 'Active carts: discard EDIT-* carts', 'feature'),
    ('feature.role_management', 'Manage roles and page visibility from app UI', 'feature'),
]


def _all_nav_codenames() -> set[str]:
    return {c for c, _, _ in ACCESS_PERMISSION_SEED}


def permissions_from_django_groups(user_groups: list[str], user: User | None) -> set[str]:
    """
    SaaS mode: do not derive app permissions from static Django group names.
    Permissions should come from Role/UserStoreRole assignments.
    Keep staff/superuser fallback for platform operations.
    """
    if user and (user.is_superuser or user.is_staff):
        return _all_nav_codenames() | {c for c, _, _ in FEATURE_PERMISSION_SEED}
    # Legacy fallback: keep existing group-based behavior for non-migrated users.
    groups = {g.lower() for g in (user_groups or [])}
    perms: set[str] = set()
    has_admin = any('admin' in g for g in groups)
    has_retail = 'retail' in groups
    has_retail_admin = 'retailadmin' in groups

    if has_admin:
        perms.update({'feature.ledger_admin', 'feature.store_management', 'nav.payments'})
    if has_retail_admin:
        perms.update({'feature.store_management', 'nav.payments'})
    if has_retail:
        perms.add('nav.payments')

    return perms


def merge_store_role_permissions(user: User, base: set[str]) -> set[str]:
    """Union Role.permissions from UserStoreRole rows (respecting assigned_stores)."""
    from backend.core.models import UserStoreRole

    qs = UserStoreRole.objects.filter(user_id=user.pk).select_related('role').prefetch_related(
        'role__permissions'
    )
    if user.assigned_stores.exists():
        allowed = set(user.assigned_stores.values_list('id', flat=True))
        qs = qs.filter(store_id__in=allowed)

    extra: set[str] = set()
    for usr in qs:
        extra.update(p.codename for p in usr.role.permissions.all())
    return base | extra
